### PR TITLE
tackle-77-update-assessment

### PR DESCRIPTION
**Pull request by jonathanvila**
### Features covered
* New endpoint PATH /assessments/{assessmentId} that will Patially or Totally UPDATE an existing assessment

### Pre steps Minikube: 
1. Install minikube : https://minikube.sigs.k8s.io/docs/start/
1. Start minikube : minikube start --kubernetes-version=v1.20.2 --cpus 4 --memory 12000
1. Enable ingress addons : minikube addons enable ingress
1. Create tackle namespace
```Shell
kubectl create namespace tackle
``` 
1. Deploy tackle ( to have the keycloak instance running ) on your K8s cluster
```Shell
$ kubectl apply -f  https://raw.githubusercontent.com/konveyor/tackle/main/kubernetes/kubernetes-tackle.yaml -n tackle
```
5. Edit the Ingress object to allow direct request to Pathfinder API
```Shell
$ kubectl edit ingress tackle -n tackle
```
Add this below `paths:`
```Yaml
     - backend:
          service:
            name: tackle-pathfinder
            port:
              number: 8080 
        path: /pathfinder
        pathType: ImplementationSpecific
```
6. Build and push 
```Shell
$ ./mvnw -U -B package -Dquarkus.container-image.push=true -Dquarkus.container-image.group={your quay user} -Dquarkus.container-image.registry=quay.io -Dquarkus.container-image.username={your quay user} -Dquarkus.container-image.password={your quay pwd} -Pnative
```
7. Change the image in the deployment tackle-pathfinder to be the one you have pushed
8. Rollout to update the image
```Shell
$ kubectl rollout restart deployment tackle-pathfinder -n tackle
```

### Test case with Minikube : Do the full API test
```Shell
$ .github/scripts/check_api.sh
```
> Expect 
```Shell
+++++ API CHECK SUCCESSFUL ++++++
```

### Pre steps non containerised local test:
1. run keycloak
```Shell
podman run -it --name keycloak --rm \     
            -e KEYCLOAK_USER=admin -e KEYCLOAK_PASSWORD=admin -e KEYCLOAK_IMPORT=/tmp/keycloak/quarkus-realm.json \
            -e DB_VENDOR=h2 -p 8180:8080 -p 8543:8443 -v ./src/main/resources/keycloak:/tmp/keycloak:Z \
            jboss/keycloak:12.0.2
```
1. run postgresql
```Shell
podman run -it \                          
            --name postgres-pathfinder -e POSTGRES_USER=pathfinder \                                               
            -e POSTGRES_PASSWORD=pathfinder -e POSTGRES_DB=pathfinder_db \                              
            -p 5433:5432 postgres:10.6
```
1. run application
```Shell
./mvnw quarkus:dev
```
 
### Test case with local non containerised : Do the full API test
```Shell
$ .github/scripts/check_api.sh localhost:8085 localhost:8180
```
> Expect 
```Shell
+++++ API CHECK SUCCESSFUL ++++++
```

### CODE REVIEW

File: src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java [hunk 1]
```diff
@@ -118,9 +114,76 @@ private Questionnaire defaultQuestionnaire() {
...
+            // Delete existing stakeholdergroups not included in current array
+            assessment.stakeholdergroups.forEach(stakegroup -> {
+                if (!assessmentDto.getStakeholderGroups().contains(stakegroup.stakeholdergroupId)) {
+                    log.info("Deleted stakegroup : " + stakegroup.stakeholdergroupId);
+                    stakegroup.delete();
+                }
+            });
+            // Add not existing stakeholdergroups included in the current array
+            assessmentDto.getStakeholderGroups().forEach(e -> {
+                log.info("Considering Stakeholdergroup : " + e);
```
Comment (m-brophy) [inline src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java:135]: this method's log calls would be better as debug level

File: src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java [hunk 2]
```diff
@@ -118,9 +114,76 @@ private Questionnaire defaultQuestionnaire() {
...
+                    AssessmentStakeholder.builder()
+                        .assessment(assessment)
+                        .stakeholderId(e)
+                        .build().persist();
+                }
+            });
+        }
+
+        assessmentDto.getQuestionnaire().getCategories().forEach(categ -> {
+            AssessmentCategory category = AssessmentCategory.find("assessment_questionnaire_id=?1 and id=?2", assessment.assessmentQuestionnaire.id, categ.getId()).<AssessmentCategory>firstResultOptional().orElseThrow(BadRequestException::new);
```
Comment (m-brophy) [inline src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java:167]: check that questionnaire/category combination in Dto are attached to the correct assessment

File: src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java [hunk 3]
```diff
@@ -118,9 +114,76 @@ private Questionnaire defaultQuestionnaire() {
...
+            }
+
+            categ.getQuestions().forEach(que -> {
+                AssessmentQuestion question = AssessmentQuestion.find("assessment_category_id=?1 and id=?2", categ.getId(), que.getId()).<AssessmentQuestion>firstResultOptional().orElseThrow(BadRequestException::new);
+
+                que.getOptions().forEach(opt -> {
+                    AssessmentSingleOption option = AssessmentSingleOption.find("assessment_question_id=?1 and id=?2", question.id, opt.getId()).<AssessmentSingleOption>firstResultOptional().orElseThrow(BadRequestException::new);
+                    if (opt.getChecked() != null) {
+                      option.selected = opt.getChecked();
+                        log.info("Setting category comment : " + category.comment);
```
Comment (m-brophy) [inline src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java:180]: this log message should be logging the checked field passed in

File: src/test/java/io/tackle/pathfinder/services/AssessmentSvcTest.java [hunk 4]
```diff
@@ -17,6 +23,7 @@
 import io.tackle.pathfinder.model.questionnaire.Questionnaire;
 import io.tackle.pathfinder.model.questionnaire.SingleOption;
 import lombok.extern.java.Log;
+import org.apache.commons.lang3.tuple.Pair;
```
Comment (m-brophy) [inline src/test/java/io/tackle/pathfinder/services/AssessmentSvcTest.java:26]: this import isn't  needed

File: src/test/java/io/tackle/pathfinder/services/AssessmentSvcTest.java [hunk 5]
```diff
@@ -93,6 +103,105 @@ public void given_Questionnaire_when_CopyQuestionnaireIntoAssessment_should_BeId
...
+        Assessment assessment = createAssessment(Questionnaire.findAll().firstResult(), 3410L);
+
+        assertThat(assessment.stakeholders).extracting(e -> e.stakeholderId).containsExactlyInAnyOrder(100L, 200L, 300L);
+        assertThat(assessment.stakeholdergroups).extracting(e -> e.stakeholdergroupId).containsExactlyInAnyOrder(500L, 600L);
+
+        AssessmentDto assessmentDto = assessmentMapper.assessmentToAssessmentDto(assessment);
+        assertThat(assessmentDto.getStakeholderGroups()).hasSize(2);
+        assertThat(assessmentDto.getStakeholders()).hasSize(3); 
+
+        // Stakeholders and Stakeholdergroups NOT send will imply leave what is there without touching it
```
Comment (m-brophy) [inline src/test/java/io/tackle/pathfinder/services/AssessmentSvcTest.java:178]: update comment to "send empty implies ..."etc

File: src/main/java/io/tackle/pathfinder/controllers/AssessmentsResource.java [hunk 6]
```diff
@@ -49,6 +50,16 @@ public Response createAssessment(@NotNull @Valid ApplicationDto data) {
   @Produces("application/json")
   public AssessmentDto getAssessment(@NotNull @PathParam("assessmentId") Long assessmentId) {
     return service.getAssessmentDtoByAssessmentId(assessmentId);
+  }  
+  
+  @PATCH
+  @Path("{assessmentId}")
+  @Produces("application/json")
+  @Consumes("application/json")
+  public AssessmentHeaderDto updateAssessment(@NotNull @PathParam("assessmentId") Long assessmentId, @NotNull @Valid AssessmentDto assessment) {
```
[the remark this answers is not in the document]
Comment (jonathanvila) [reply src/main/java/io/tackle/pathfinder/controllers/AssessmentsResource.java:59]: After conversation with @carlosthe19916, seems a better approach to consider the Pathparam as the source of truth and not consider at all the assessmentId that could come in the body.

File: src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java [hunk 7]
```diff
@@ -4,11 +4,7 @@
 import io.tackle.pathfinder.dto.AssessmentHeaderDto;
 import io.tackle.pathfinder.dto.AssessmentStatus;
 import io.tackle.pathfinder.mapper.AssessmentMapper;
-import io.tackle.pathfinder.model.assessment.Assessment;
-import io.tackle.pathfinder.model.assessment.AssessmentCategory;
-import io.tackle.pathfinder.model.assessment.AssessmentQuestion;
-import io.tackle.pathfinder.model.assessment.AssessmentQuestionnaire;
-import io.tackle.pathfinder.model.assessment.AssessmentSingleOption;
+import io.tackle.pathfinder.model.assessment.*;
```
[the remark this answers is not in the document]
Comment (jonathanvila) [reply src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java:7]: :+1:

File: src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java (hunk 1 above)
Comment (jonathanvila) [reply src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java:135]: :+1:

File: src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java (hunk 2 above)
Comment (jonathanvila) [reply src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java:167]: :+1:

File: src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java (hunk 3 above)
Comment (jonathanvila) [reply src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java:180]: :+1:

File: src/test/java/io/tackle/pathfinder/services/AssessmentSvcTest.java (hunk 5 above)
Comment (jonathanvila) [reply src/test/java/io/tackle/pathfinder/services/AssessmentSvcTest.java:178]: :+1:

**Review by m-brophy — APPROVED** [review]

**Pull request merged**